crypto/kyber/p_kyber_test.cc

```diff
@@ -4,280 +4,463 @@
+    size_t pk_bytes = kyber_public_key_size(pkey_id);
+    size_t sk_bytes = kyber_secret_key_size(pkey_id);
 
-  EVP_PKEY_CTX_free(kyber_pkey_ctx);
-}
+    EXPECT_EQ(pk_bytes + sk_bytes, (size_t) EVP_PKEY_size(kyber_pkey));
+    EXPECT_EQ(8 * (pk_bytes + sk_bytes), (size_t) EVP_PKEY_bits(kyber_pkey));
 
-TEST(Kyber512Test, KEMOperations) {
-  // Basic functional test for KYBER512.
-  // Simulate two sides of the key exchange mechanism.
-  size_t shared_secret_len = KYBER512_SHARED_SECRET_BYTES;
-  size_t ciphertext_len = KYBER512_CIPHERTEXT_BYTES;
-  uint8_t shared_secret_alice[KYBER512_SHARED_SECRET_BYTES];
-  uint8_t shared_secret_bob[KYBER512_SHARED_SECRET_BYTES];
-  uint8_t ciphertext_alice[KYBER512_CIPHERTEXT_BYTES];
-  uint8_t ciphertext_bob[KYBER512_CIPHERTEXT_BYTES];
-
-  // Alice generates the key pair.
-  EVP_PKEY_CTX *kyber_pkey_ctx_alice = EVP_PKEY_CTX_new_id(EVP_PKEY_KYBER512, nullptr);
-  EVP_PKEY *kyber_pkey_alice = EVP_PKEY_new();
-  EXPECT_TRUE(EVP_PKEY_keygen_init(kyber_pkey_ctx_alice));
-  EXPECT_TRUE(EVP_PKEY_keygen(kyber_pkey_ctx_alice, &kyber_pkey_alice));
-
-  // Alice passes the public key to Bob.
-  const KYBER512_KEY *kyber_key_alice = (KYBER512_KEY *)(kyber_pkey_alice->pkey.ptr);
-
-  EVP_PKEY_CTX *kyber_pkey_ctx_bob = EVP_PKEY_CTX_new_id(EVP_PKEY_KYBER512, nullptr);
-  EVP_PKEY *kyber_pkey_bob = EVP_PKEY_new_raw_public_key(EVP_PKEY_KYBER512,
-                                                         NULL,
-                                                         kyber_key_alice->pub, /* this method performs a memcpy internally */
-                                                         KYBER512_PUBLIC_KEY_BYTES);
-  kyber_pkey_ctx_bob->pkey = kyber_pkey_bob;
-
-  // Bob generates a shared secret and encapsulates it.
-  ASSERT_TRUE(EVP_PKEY_encapsulate(kyber_pkey_ctx_bob, ciphertext_bob, &ciphertext_len, shared_secret_bob, &shared_secret_len));
-
-  // Bob sends the ciphertext to Alice.
-  OPENSSL_memcpy(ciphertext_alice, ciphertext_bob, ciphertext_len);
-
-  // Alice decapsulates the ciphertext to obtain the shared secret.
-  ASSERT_TRUE(EVP_PKEY_decapsulate(kyber_pkey_ctx_alice, shared_secret_alice, &shared_secret_len, ciphertext_alice, ciphertext_len));
-
-  // Verify that Alice and Bob have the same shared secret.
-  for (size_t i = 0; i < shared_secret_len; i++) {
-    EXPECT_EQ(shared_secret_alice[i], shared_secret_bob[i]);
+    EVP_PKEY_CTX_free(kyber_pkey_ctx);
   }
+}
 
-  // Invalidate the ciphertext and run decapsulate on it.
-  ciphertext_alice[0] ^= 1;
+TEST(KyberTest, KEMOperations) {
+  for (auto pkey_id : pkeyIDsToTest) {
+    // Basic functional test for Kyber that simulates two sides
+    // of the key exchange mechanism.
+    size_t shared_secret_len = kyber_shared_secret_size(pkey_id);
+    size_t ciphertext_len = kyber_ciphertext_size(pkey_id);
+    size_t public_key_len = kyber_public_key_size(pkey_id);
```

Comment:
Why not call EVP_PKEY_encapsulate/EVP_PKEY_decapsulate with null to get these sizes instead of having to make a custom test mapping?